[GTK][WPE] Use a WorkQueue instead of CompositingRunLoop in ThreadedCompositor

carlosgcampos opened the pull request:
#### 20c7683a6fb0b4b600723221131c3e2309e4a096
<pre>
[GTK][WPE] Use a WorkQueue instead of CompositingRunLoop in ThreadedCompositor
<a href="https://bugs.webkit.org/show_bug.cgi?id=302638">https://bugs.webkit.org/show_bug.cgi?id=302638</a>

Reviewed by Nikolas Zimmermann.

It&apos;s simpler to use a WorkQueue and move the state handling to
ThreadedCompositor which also improves the readability.

Covered by existing tests since there&apos;s no change in behavior.

Canonical link: <a href="https://commits.webkit.org/303179@main">https://commits.webkit.org/303179@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/358da295c3f1aba82c59eb80b3c9a229df65458e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131402 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3732 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42366 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138917 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83176 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3754 "Built successfully") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3574 "Hash 358da295 for PR 54046 does not build (failure)") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100353 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/67896 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134348 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2709 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117619 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81134 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2626 "Passed tests") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82100 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/123424 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111235 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35741 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141543 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/129856 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3477 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36262 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108705 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3523 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/156/builds/3574 "Hash 358da295 for PR 54046 does not build (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108921 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2635 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113949 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/56654 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20447 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3539 "Built successfully") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32370 "Found 1 new test failure: http/tests/workers/service/registerServiceWorkerClient-after-network-process-crash.html (failure)") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/162873 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3361 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66947 "Built successfully") | [⏳ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3561 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3469 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
```diff
@@ -65,11 +64,12 @@ Ref<ThreadedCompositor> ThreadedCompositor::create(LayerTreeHost& layerTreeHost)
     , m_sceneState(&m_layerTreeHost->sceneState())
     , m_flipY(m_surface->shouldPaintMirrored())
-    , m_compositingRunLoop(makeUnique<CompositingRunLoop>([this] { renderLayerTree(); }))
+    , m_renderTimer(m_workQueue->runLoop(), "ThreadedCompositor::RenderTimer"_s, this, &ThreadedCompositor::renderLayerTree)
```
How about naming it renderSceneTimer and renderLayerTree -> renderScene?

carlosgcampos replied:
We can do renames in follow up commits.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
```diff
@@ -281,14 +289,19 @@ void ThreadedCompositor::paintToCurrentGLContext(const TransformationMatrix& mat
 
+    {
+        Locker locker { m_state.lock };
+        m_state.state = State::InProgress;
```
So if it's suspended we just retain the previous state?

carlosgcampos replied:
Yes, no change in behavior here, if composition is requested during the suspension we need to handle that on resume.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
```diff
@@ -350,24 +363,46 @@ uint32_t ThreadedCompositor::requestComposition()
+    switch (m_state.state) {
+    case State::Idle:
+        m_state.state = State::Scheduled;
+        if (!m_suspendedCount.load())
```
Why atomic if this is protected by the state lock?

carlosgcampos replied:
Because suspended is not really part of the state, we don't take the state lock in suspend/resume, except for the resume case but only when it's 0.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
```diff
@@ -350,24 +363,46 @@ uint32_t ThreadedCompositor::requestComposition()
 }
 
-RunLoop* ThreadedCompositor::runLoop()
+void ThreadedCompositor::frameComplete()
```
I dislike this name - whatever we choose should be consistent with the operation that is performed, currently this is the result of renderLayerTree...

carlosgcampos replied:
I agree, but again, out of scope of this PR, we can decide better names for everything in a different PR.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
```diff
@@ -420,8 +455,7 @@ void ThreadedCompositor::updateFPSCounter()
 {
-    ASSERT(m_compositingRunLoop);
-    m_compositingRunLoop->performTask([protectedThis = Ref { *this }, info = WTFMove(info), completionHandler = WTFMove(completionHandler)]() mutable {
+    m_workQueue->dispatchSync([protectedThis = Ref { *this }, info = WTFMove(info), completionHandler = WTFMove(completionHandler)]() mutable {
```
Why do we need to protect this precisely here and not in the constructor?

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:
```diff
@@ -117,14 +118,25 @@ class ThreadedCompositor : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPt
         float deviceScaleFactor { 1 };
     } m_attributes;
 
+    RunLoop::Timer m_renderTimer;
```
That could be part of the State struct?

carlosgcampos replied:
Why? It's the render timer and not protected by the state lock.

nikolaszimmermann replied:
It logically belongs to the State - the renderTimer starts the state transitions. I have no strong opinion though - I only thought about "grouping" it, not that it would be protected by the lock, but I see this could be confusing as well.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
```diff
@@ -350,24 +363,46 @@ uint32_t ThreadedCompositor::requestComposition()
+        break;
+    case State::ScheduledWhileInProgress:
+        m_state.state = State::Scheduled;
+        if (!m_suspendedCount.load())
```
Isn't this already protected?

carlosgcampos replied:
What do you mean?

nikolaszimmermann replied:
I meant using the state lock, but I see we don't take the lock, when we mutate it elsewhere.

carlosgcampos committed:
[GTK][WPE] Use a WorkQueue instead of CompositingRunLoop in ThreadedCompositor
https://bugs.webkit.org/show_bug.cgi?id=302638

Reviewed by Nikolas Zimmermann.

It's simpler to use a WorkQueue and move the state handling to
ThreadedCompositor which also improves the readability.

Covered by existing tests since there's no change in behavior.

Canonical link: https://commits.webkit.org/303179@main
